### PR TITLE
Add tests to verify assembler output -- Fix DoNotOptimize.

### PR DESCRIPTION
**Pull request by EricWF**
For things like `DoNotOptimize`, `ClobberMemory`, and even `KeepRunning()`,
it is important exactly what assembly they generate. However, we currently
have no way to test this. Instead it must be manually validated every
time a change occurs -- including a change in compiler version.

This patch attempts to introduce a way to test the assembled output automatically.
It's mirrors how LLVM verifies compiler output, and it uses LLVM FileCheck to run
the tests in a similar way.

The tests function by generating the assembly for a test in CMake, and then
using FileCheck to verify the // CHECK lines in the source file are found
in the generated assembly.

Currently, the tests only run on 64-bit x86 systems under GCC and Clang,
and when FileCheck is found on the system.

Additionally, this patch tries to improve the code gen from DoNotOptimize. This should probably be a separate change, but I needed something to test.

This PR is largely just to run the new patch against the bots to ensure nothing has broken.

### CODE REVIEW

**Comment by dominichamon** [comment]
Can you add a markdown documentation file about AssemblyTests or whatever name you choose, essentially with the contents of your commit message. Maybe more detail.

And then cross-link to it from the README?

**Comment by pleroy** [comment]
TIL about LLVM FileCheck.  Pretty cool!

**Comment by EricWF** [comment]
@dominichamon Yep. Can do.

@pleroy:
> TIL about LLVM FileCheck. Pretty cool!

Indeed. It's an amazing tool to have when you need to test string output.

**Comment by EricWF** [comment]
@dominichamon I added documentation as requested. Please take a look at your leisure.

File: docs/AssemblyTests.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
+# Assembly Tests
+
+The Benchmark library provides a number of functions who's primary
```
Comment (dominichamon) [inline docs/AssemblyTests.md]: 'whose'

File: docs/AssemblyTests.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    return ExternInt + 1;
+
+    // CHECK: movl ExternInt(%rip), %eax
+    // CHECK: addl %eax
+    // CHECK: ret
+}
+
+```
+
+#### LLVM Filecheck
```
Comment (dominichamon) [inline docs/AssemblyTests.md]: this should be ### i think, as you're only at ## above.

Comment (EricWF) [reply docs/AssemblyTests.md]: If it's no matter, I actually prefer having the smaller section headers for these bits. IMHO it's looks nicer and flows better.

**Comment by EricWF** [comment]
@dominichamon Any final words?

**Comment by dominichamon** [comment]
I'd prefer if the bazel TODO linked to a github issue so we can track it, but otherwise, Ship It.

**Pull request merged**